<a href="https://colab.research.google.com/github/Rizki-Code156/pm-turi2-preprocessing-RizkiAidilFazri/blob/main/PM_P4_Rizki%20Aidil%20Fazri_2488010021.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

#C. Langkah Praktikum


##Membuat Dataset Karyawan


In [69]:
import pandas as pd
import numpy as np

data = {
    'usia': [23, 30, np.nan, 35, 28, 42, 31, np.nan, 26, 38],
    'pendapatan': [4000000, 5500000, 4500000, np.nan, 5000000,
                   7500000, 6000000, 4800000, np.nan, 7000000],
    'pendidikan': ['SMA', 'S1', 'S1', 'S2', 'SMA',
                   'S2', 'S1', 'SMA', 'S1', 'S2'],
    'kota': ['Cirebon', 'Bandung', 'Jakarta', 'Cirebon', 'Bandung',
             'Jakarta', 'Cirebon', 'Bandung', 'Jakarta', 'Cirebon'],
    'membeli': ['Tidak', 'Ya', 'Tidak', 'Ya', 'Tidak',
                'Ya', 'Ya', 'Tidak', 'Tidak', 'Ya']
}

df = pd.DataFrame(data)

df

,usia,pendapatan,pendidikan,kota,membeli
0,23.0,4000000.0,SMA,Cirebon,Tidak
1,30.0,5500000.0,S1,Bandung,Ya
2,NaN,4500000.0,S1,Jakarta,Tidak
3,35.0,NaN,S2,Cirebon,Ya
4,28.0,5000000.0,SMA,Bandung,Tidak
5,42.0,7500000.0,S2,Jakarta,Ya
6,31.0,6000000.0,S1,Cirebon,Ya
7,NaN,4800000.0,SMA,Bandung,Tidak
8,26.0,NaN,S1,Jakarta,Tidak
9,38.0,7000000.0,S2,Cirebon,Ya


##Langkah 1 — Menangani nilai hilang (imputasi median/modus).

In [70]:
df['usia'] = df['usia'].fillna(df['usia'].median())
df['pendapatan'] = df['pendapatan'].fillna(df['pendapatan'].median())

##Langkah 2 — Memisahkan fitur (X) dan label (y).

In [71]:
X = df.drop(columns=['membeli'])
y = df['membeli']

##Langkah 3 — Encoding kategorikal (ordinal & one-hot).

In [72]:
X['pendidikan'] = X['pendidikan'].map({'SMA':0,'S1':1,'S2':2}) # ordinal
X = pd.get_dummies(X, columns=['kota'], dtype=int) # nominal
y = y.map({'Tidak':0,'Ya':1})

##Langkah 4 — Membagi data (SEBELUM scaling).

In [73]:
from sklearn.model_selection import train_test_split
X_train,X_test,y_train,y_test = train_test_split(
X, y, test_size=0.3, random_state=42)

##Langkah 5 — Penskalaan (fit pada latih, transform pada latih & uji).

In [74]:
from sklearn.preprocessing import StandardScaler
num = ['usia','pendapatan','pendidikan']
sc = StandardScaler()
X_train[num] = sc.fit_transform(X_train[num])
X_test[num] = sc.transform(X_test[num]) # cegah data leakage

##Langkah 6 — Verifikasi: tidak ada nilai hilang, semua numerik, skala seragam.

In [75]:
print("Jumlah data training:", len(X_train))
print("Jumlah data testing:", len(X_test))
print("\nNilai yang hilang:")
print(X.isna().sum())
print("\nData X_train:")
display(X_train)

Jumlah data training: 7
Jumlah data testing: 3

Nilai yang hilang:
usia            0
pendapatan      0
pendidikan      0
kota_Bandung    0
kota_Cirebon    0
kota_Jakarta    0
dtype: int64

Data X_train:


,usia,pendapatan,pendidikan,kota_Bandung,kota_Cirebon,kota_Jakarta
0,-1.767311,-1.318274,-1.028992,0,1,0
7,-0.080332,-0.454843,-1.028992,1,0,0
2,-0.080332,-0.778630,0.171499,0,0,1
9,1.606646,1.919592,1.371989,0,1,0
4,-0.642658,-0.238985,-1.028992,1,0,0
3,0.931855,0.030837,1.371989,0,1,0
6,0.032133,0.840303,0.171499,0,1,0


#D. Latihan Mandiri

##1. Ganti imputasi usia menjadi mean; bandingkan hasilnya.

In [76]:
df_mean = pd.DataFrame({
    'usia': [23, 30, np.nan, 35, 28, 42, 31, np.nan, 26, 38]
})

mean_usia = df_mean['usia'].mean()

print("Mean usia:", mean_usia)

df_mean['usia'] = df_mean['usia'].fillna(mean_usia)

print("\nData usia setelah imputasi mean:")
print(df_mean)

Mean usia: 31.625

Data usia setelah imputasi mean:
     usia
0  23.000
1  30.000
2  31.625
3  35.000
4  28.000
5  42.000
6  31.000
7  31.625
8  26.000
9  38.000


Setelah menggunakan metode imputasi mean, nilai rata-rata usia adalah 31.625. Nilai usia yang sebelumnya kosong diganti dengan nilai mean tersebut.

##2. Gunakan MinMaxScaler (0–1) sebagai ganti StandardScaler; tampilkan describe().

In [77]:
from sklearn.preprocessing import MinMaxScaler

sc_minmax = MinMaxScaler()

X_train_minmax = X_train.copy()
X_test_minmax = X_test.copy()

X_train_minmax[num] = sc_minmax.fit_transform(X_train_minmax[num])
X_test_minmax[num] = sc_minmax.transform(X_test_minmax[num])

print("Hasil MinMaxScaler:")
display(X_train_minmax[num].describe())

Hasil MinMaxScaler:


,usia,pendapatan,pendidikan
count,7.000000,7.000000,7.000000
mean,0.523810,0.407143,0.428571
std,0.320136,0.333591,0.449868
min,0.000000,0.000000,0.000000
25%,0.416667,0.216667,0.000000
50%,0.500000,0.333333,0.500000
75%,0.666667,0.541667,0.750000
max,1.000000,1.000000,1.000000


MinMaxScaler digunakan untuk mengubah skala fitur numerik ke rentang 0 sampai 1. Berdasarkan hasil describe(), nilai minimum pada fitur usia, pendapatan, dan pendidikan adalah 0, sedangkan nilai maksimum adalah 1. Hal ini menunjukkan bahwa proses penskalaan berhasil dilakukan.

##3. Tambahkan kolom kategorikal baru dan terapkan one-hot encoding.

In [78]:
df_status = df.copy()

# Menambahkan kolom kategorikal baru
df_status['status'] = [
    'Belum Menikah',
    'Menikah',
    'Belum Menikah',
    'Menikah',
    'Belum Menikah',
    'Menikah',
    'Menikah',
    'Belum Menikah',
    'Belum Menikah',
    'Menikah'
]

# Melakukan One-Hot Encoding
df_status = pd.get_dummies(
    df_status,
    columns=['status'],
    dtype=int
)

# Menampilkan hasil
print("Data setelah menambahkan kolom status dan One-Hot Encoding:")
display(df_status)

Data setelah menambahkan kolom status dan One-Hot Encoding:


,usia,pendapatan,pendidikan,kota,membeli,status_Belum Menikah,status_Menikah
0,23.0,4000000.0,SMA,Cirebon,Tidak,1,0
1,30.0,5500000.0,S1,Bandung,Ya,0,1
2,30.5,4500000.0,S1,Jakarta,Tidak,1,0
3,35.0,5250000.0,S2,Cirebon,Ya,0,1
4,28.0,5000000.0,SMA,Bandung,Tidak,1,0
5,42.0,7500000.0,S2,Jakarta,Ya,0,1
6,31.0,6000000.0,S1,Cirebon,Ya,0,1
7,30.5,4800000.0,SMA,Bandung,Tidak,1,0
8,26.0,5250000.0,S1,Jakarta,Tidak,1,0
9,38.0,7000000.0,S2,Cirebon,Ya,0,1


Kolom kategorikal baru status berhasil ditambahkan ke dataset. Selanjutnya, kolom status diubah menggunakan one-hot encoding sehingga kategori Menikah dan Belum Menikah direpresentasikan dalam bentuk kolom numerik 0 dan 1. Dengan demikian, data kategorikal dapat digunakan dalam proses pembelajaran mesin.

##4. Jelaskan mengapa fit_transform pada latih tetapi hanya transform pada uji.

fit_transform digunakan pada data latih karena scaler perlu mempelajari parameter dari data latih terlebih dahulu, kemudian melakukan transformasi terhadap data tersebut.

Sedangkan pada data uji hanya digunakan transform karena data uji harus menggunakan parameter yang sudah diperoleh dari data latih. Hal ini dilakukan untuk mencegah data leakage, yaitu informasi dari data uji ikut memengaruhi proses preprocessing.

#E. Refleksi

##1. Mengapa urutan 'split dulu, baru scaling' penting?

Split data dilakukan terlebih dahulu sebelum scaling untuk mencegah data leakage. Jika scaling dilakukan sebelum data dibagi menjadi data training dan testing, informasi dari data testing dapat ikut digunakan dalam proses scaling. Dengan melakukan split terlebih dahulu, parameter scaling dipelajari hanya dari data training, kemudian diterapkan pada data testing.

##2. Kapan label encoding dan kapan one-hot encoding?

Label encoding digunakan untuk data kategorikal yang memiliki urutan atau tingkatan, seperti pendidikan SMA, S1, dan S2. Sedangkan one-hot encoding digunakan untuk data kategorikal yang tidak memiliki urutan, seperti nama kota Cirebon, Bandung, dan Jakarta.

##3. Apa perbedaan normalisasi dan standardisasi?

Normalisasi mengubah nilai data ke dalam rentang tertentu, misalnya 0 sampai 1 seperti yang dilakukan oleh MinMaxScaler. Sedangkan standardisasi mengubah data berdasarkan nilai rata-rata dan standar deviasi sehingga data umumnya memiliki rata-rata 0 dan standar deviasi 1 seperti pada StandardScaler.